lib/galaxy/tools/cache.py

```diff
@@ -303,10 +303,7 @@ def rebuild(self):
         try:
             session = self.app.install_model._SessionLocal()
             self.repositories = session.query(ToolShedRepository).options(
-                defer(ToolShedRepository.metadata),
-                joinedload('tool_dependencies').subqueryload('tool_shed_repository').options(
-                    defer(ToolShedRepository.metadata)
-                ),
+                defer(ToolShedRepository.metadata), joinedload('tool_dependencies')
```

Comment:
The joinedload you're doing here does not look equivalent, but when I added this initially we were using the global sessionmaker, meaning that under some circumstances flushing in some unrelated code would require refetching attributes when accessing repository attributes, which failed because the session was closed. By using a session that is completely separate I think we don't need to worry about that anymore.

So what we can do is bind the session to the ToolShedRepositoryCache instance, and only close it when shutting down the app (or even let it get garbage collected, not sure we really need to explicitly close).

Reply:
Unless I am mistaken, the simplified query is fetching the same records, respecting the same deferred option. Here's a simplified model: https://gist.github.com/ic4f/765bc288b4282a34ed383e5b23e8fd0a - if you run this, you'll see that while the generated SQL is very different (see SQL in comments edited for readability), it fetches the same data. The non-empty `callables` is the deferred `metadata`.
I may be missing something, but it seems to me the subquery load was redundant. (this also exposed a bug in SA 1.4, now fixed: https://github.com/sqlalchemy/sqlalchemy/issues/6221)

As for a separate session - I don't think I know enough about this context, but if we need `ToolShedRepositoryCache` attributes to never expire (just be available  once loaded at startup?), then a separate session seems to be a good option, I think.  Are you considering this modification for this PR? (If the no-subquery load works correctly, I'd rather leave it as is and change the session handling in its own PR (as it is technically not required for moving to 1.4).

Reply:
Yeah, I think you're right, my mental model of the repository <-> tool_dependency relationship was that you can have repository a <-> tool_dependency a -> repository b, but I think that is implemented with RepositoryRepositoryDependencyAssociation.

This probably never worked as I was hoping it would ... I couldn't reproduce the bug this was supposed to fix, so that's not a surprise.

Reply:
Yes, this - `repository a <-> tool_dependency a -> repository b` is what I assumed it was expected to do, but I came to the same conclusion: that would be done via `RepositoryRepositoryDependencyAssociation`.